include/onnxruntime/core/framework/tensor_shape.h

```diff
@@ -134,14 +121,22 @@ class TensorShape : private std::vector<int64_t> {
      empty shape or 1D shape (1) is regarded as scalar tensor
   */
   bool IsScalar() const {
-    size_t len = size();
-    return len == 0 || (len == 1 && operator[](0) == 1);
+    size_t len = values_.size();
+    return len == 0 || (len == 1 && values_[0] == 1);
   }
 
-  static const TensorShape& ReinterpretBaseType(const std::vector<int64_t>& dimensions) {
-    static_assert(sizeof(TensorShape) == sizeof(std::vector<int64_t>), "Size of TensorShape prevents safe casting from vector");
-    return *static_cast<const TensorShape*>(&dimensions);
-  }
+ private:
+
+  struct External {};
+  TensorShape(External, gsl::span<int64_t> buffer) : values_{buffer} {}
+
+  void Allocate(size_t size);
+
+  gsl::span<int64_t> values_;
+  int64_t small_buffer_[5];
+  std::unique_ptr<int64_t[]> allocated_buffer_;
```

Comment:
I feel that we could use std::pmr::vector with pool allocator sitting on top of the buffer with upstream new_delete. That way we perfectly simulate small value optimization with first bytes being taking from the inline buffer and bigger buffers from heap. This way we would get rid of unique_ptr and all the logic associated with it.

Reply:
Furthermore, we could abstract all the logic of manipulating mutable shapes inside TensorShape based on pmr and enjoy correctness and the absence of heap.

Reply:
We talked about it outside of github, here's a summary:

std::pmr::vector would be useful in all of the code that currently uses TensorShape::GetDimsAsVector to take advantage of a small size optimized vector. As TensorShape isn't a vector (no runtime adding/removing of elements) it's not needed there (and would take up more space).

Sadly there's no way to remove the unique_ptr member as without it there is no way to know if the memory is the small block, allocated memory, or external memory from TensorShape::FromExistingBuffer.